File: Algorithms/002.AddTwoNumbers.py
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        dummyHead = ListNode(0)
        current = dummyHead
        carry = 0
        while l1 or l2:
            x = l1.val if l1 else 0
            y = l2.val if l2 else 0
            digit = x + y + carry
            carry = digit / 10
            current.next = ListNode(digit % 10)
            current = current.next
            if l1: l1 = l1.next
            if l2: l2 = l2.next
        if carry:
            current.next = ListNode(carry)
        return dummyHead.next

File: Algorithms/cpp/002.AddTwoNumbers.cpp
class Solution {
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
        ListNode *dummyHead = new ListNode(0);
        ListNode *current = dummyHead;
        int carry = 0;
        while (l1 != NULL || l2 != NULL) {
            int x = l1 != NULL? l1->val : 0;
            int y = l2 != NULL? l2->val : 0;
            int digit = x + y + carry;
            carry = digit / 10;
            current->next = new ListNode(digit % 10);
            current = current->next;
            if (l1 != NULL) { l1 = l1->next; }
            if (l2 != NULL) { l2 = l2->next; }
        }
        if (carry) { current->next = new ListNode(carry); }
        return dummyHead->next;
    }
};
